File: src/main/java/io/github/binaryfoo/decoders/apdu/APDUSequenceDecoder.java
package io.github.binaryfoo.decoders.apdu;

import io.github.binaryfoo.DecodedData;
import io.github.binaryfoo.Decoder;
import io.github.binaryfoo.HexDumpFactory;
import io.github.binaryfoo.decoders.DecodeSession;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class APDUSequenceDecoder implements Decoder {

    private ReplyAPDUDecoder replyDecoder;
    private CommandAPDUDecoder[] commandDecoders;
    private HexDumpFactory hexDumpFactory = new HexDumpFactory();

    public APDUSequenceDecoder(ReplyAPDUDecoder replyDecoder, CommandAPDUDecoder... commandDecoders) {
        this.replyDecoder = replyDecoder;
        this.commandDecoders = commandDecoders;
    }

    @Override
    public List<DecodedData> decode(@NotNull String input, int startIndexInBytes, @NotNull DecodeSession session) {
        ArrayList<DecodedData> list = new ArrayList<DecodedData>();
        for (String line : input.split("\\s+")) {
            try {
                CommandAPDUDecoder commandDecoder = getCommandDecoder(line);
                DecodedData decoded;
                if (commandDecoder != null) {
                    session.setCurrentCommand(commandDecoder.getCommand());
                    decoded = commandDecoder.decode(line, startIndexInBytes, session);
                } else {
                    decoded = replyDecoder.decode(line, startIndexInBytes, session);
                }
                decoded.setHexDump(hexDumpFactory.splitIntoByteLengthStrings(line, startIndexInBytes));
                list.add(decoded);
                startIndexInBytes = decoded.getEndIndex();
            } catch (Exception e) {
                list.add(DecodedData.primitive(line, "Failed to decode: " + e.getMessage(), 0, 0));
            }
        }
        return list;
    }

    private CommandAPDUDecoder getCommandDecoder(String input) {
        APDUCommand command = APDUCommand.fromHex(input.substring(0, 4));
        for (CommandAPDUDecoder commandDecoder : commandDecoders) {
            if (commandDecoder.getCommand() == command) {
                return commandDecoder;
            }
        }
        return null;
    }

    @Override
    public String validate(String input) {
        return null;
    }

    @Override
    public int getMaxLength() {
        return Integer.MAX_VALUE;
    }
}

File: src/main/java/io/github/binaryfoo/decoders/apdu/APDUSequenceDecoder.kt
package io.github.binaryfoo.decoders.apdu

import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.Decoder
import io.github.binaryfoo.decoders.*
import io.github.binaryfoo.decoders.annotator.BackgroundReading
import io.github.binaryfoo.hex.HexDumpElement
import java.util.*

class APDUSequenceDecoder(private val replyDecoder: ReplyAPDUDecoder, vararg commandDecoders: CommandAPDUDecoder) : Decoder {
  private val _commandDecoders: Array<CommandAPDUDecoder> = arrayOf(*commandDecoders)
  private val signedDataRecoverers = listOf(
      IssuerPublicKeyDecoder(),
      ICCPublicKeyDecoder(),
      SignedStaticApplicationDataDecoder(),
      SignedDynamicApplicationDataDecoder()
  )

  override fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): List<DecodedData> {
    var runningStartIndexInBytes = startIndexInBytes
    val list = ArrayList<DecodedData>()
    input.toUpperCase().split(Regex("\\s+")).filter { it.isNotBlank() }.forEach { line ->
      try {
        val commandDecoder = getCommandDecoder(line)
        val decoded: DecodedData
        if (commandDecoder != null) {
          session.currentCommand = commandDecoder.getCommand()
          decoded = commandDecoder.decode(line, runningStartIndexInBytes, session)
          decoded.category = "c-apdu"
          decoded.backgroundReading = BackgroundReading.readingFor(commandDecoder.getCommand())
        } else {
          decoded = replyDecoder.decode(line, runningStartIndexInBytes, session)
          decoded.category = "r-apdu"
        }
        decoded.hexDump = HexDumpElement.splitIntoByteLengthStrings(line, runningStartIndexInBytes)
        runningStartIndexInBytes = decoded.endIndex
        list.add(decoded)
      } catch (e: Exception) {
        list.add(DecodedData.primitive(line, "Failed to decode: " + e.message, 0, 0))
      }
    }
    postProcess(list, session)
    return list
  }

  private fun getCommandDecoder(input: String): CommandAPDUDecoder? {
    val command = APDUCommand.fromHex(input.substring(0, 4))
    return _commandDecoders.firstOrNull { it.getCommand() == command }
  }

  override fun validate(input: String?): String? {
    return null
  }

  override fun getMaxLength(): Int {
    return Integer.MAX_VALUE
  }

  fun postProcess(decoded: List<DecodedData>, session: DecodeSession) {
    for (processor in signedDataRecoverers) {
      try {
        processor.decodeSignedData(session, decoded)
      } catch (e: Exception) {
      }
    }
  }
}
